src/tapedfunction.jl

```diff
@@ -1,48 +1,43 @@
-mutable struct Box{T}
-    id::Symbol
-    val::T
-end
-
-## methods for Box
-Box(x) = Box(gensym(), x)
-Box{T}(x) where {T} = Box{T}(gensym(), x)
-Base.show(io::IO, box::Box) = print(io, "Box(", box.id, ")")
-
 ## Instruction and TapedFunction
 
 abstract type AbstractInstruction end
 const RawTape = Vector{AbstractInstruction}
 
+struct TypedVar{T}
+    id::Symbol
+end
+
 """
     Instruction
 
 An `Instruction` stands for a function call
 """
-struct Instruction{F, N} <: AbstractInstruction
+struct Instruction{F, N, TO} <: AbstractInstruction
     func::F
-    input::NTuple{N, Symbol}
-    output::Symbol
+    input::NTuple{N, TypedVar{<:Any}}
+    output::TypedVar{TO}
 end
 
-struct GotoInstruction <: AbstractInstruction
-    condition::Symbol
+struct GotoInstruction{T} <: AbstractInstruction
+    condition::TypedVar{T}
     # we enusre a 1-to-1 mapping between ir.code and instruction
     # so here we can use the index directly.
     dest::Int
 end
 
-struct ReturnInstruction <: AbstractInstruction
-    arg::Symbol
+struct ReturnInstruction{T} <: AbstractInstruction
+    arg::TypedVar{T}
 end
 
 mutable struct TapedFunction{F}
     func::F # maybe a function, a constructor, or a callable object
     arity::Int
     ir::Core.CodeInfo
     tape::RawTape
+    unified_tape::Vector{FunctionWrapper{Nothing, Tuple{TapedFunction}}}
     counter::Int
-    bindings::Dict{Symbol, Box{<:Any}}
-    retval::Symbol
+    bindings::Dict{Symbol, Any}
+    retval::TypedVar
```

Comment:
```suggestion
    retval::TypedVar{T}
```
This makes `retval` a concrete type again which avoids dynamic dispatch. The original `retval::Symbol` is more performant.